detection_rules/integrations.py

```diff
@@ -0,0 +1,135 @@
+        for int_man in integration_manifests:
+            validated_manifest = IntegrationManifestSchema(unknown=EXCLUDE).load(int_man)
+            package_version = validated_manifest.pop("version")
+            final_integration_manifests[ints][package_version] = validated_manifest
+
+    manifest_file = gzip.open(manifest_file_path, "w+")
+    manifest_file_bytes = json.dumps(final_integration_manifests).encode("utf-8")
+    manifest_file.write(manifest_file_bytes)
+
+
+def find_least_compatible_version(package: str, integration: str,
+                                  current_stack_version: str, packages_manifest: dict) -> str:
+    """Finds least compatible version for specified integration based on stack version supplied."""
+    integration_manifests = packages_manifest[package]
+    least_compatible_version = None
+    compatible_versions = dict()
+
+    def compare_versions(int_ver: str, pkg_ver: str) -> bool:
+        """Compares integration and package version"""
+        pkg_major, pkg_minor = Version(pkg_ver)
+        int_major, int_minor = Version(int_ver)[:2]
+
+        if int(int_major) < int(pkg_major) or int(pkg_major) > int(int_major):
+            return(False)
+        compatible = Version(int_ver) <= Version(pkg_ver)
+        return(compatible)
+
+    for ver, manifest in integration_manifests.items():
+        kibana_compat_vers = re.sub(r"\>|\<|\=|\^", "", manifest["conditions"]["kibana.version"])
+        kibana_compat_vers = kibana_compat_vers.split(" || ")
+        bool_checks = list()
+        for kcv in kibana_compat_vers:
+            bool_checks.append(compare_versions(kcv, current_stack_version))
+        if any(bool_checks):
+            compatible_versions.setdefault(ver, True)
+
+    if len(list(compatible_versions.keys())) > 0:
+        compatible_version_list = list(compatible_versions.keys())
+        compatible_version_list.sort(key=Version)
+        least_compatible_version = compatible_version_list[0]
+    else:
+        raise Exception(f"no compatible version for integration {package}:{integration}")
+    return least_compatible_version
+
+
+def get_integration_packages(token: str, org: str, repo: str, branch: str, folder: str) -> dict:
+    """Gets integration packages object containing versioned packages and manifest content."""
+    github = GithubClient(token)
+    client = github.authenticated_client
+    organization = client.get_organization(org)
+    repository = organization.get_repo(repo)
+    sha = get_sha_for_branch(repository, branch)
+    integration_manifest = get_integration_manifests(repository, sha, folder)
+    return integration_manifest
+
+
+def get_sha_for_branch(repository, branch: str) -> str:
+    """Returns a commit PyGithub object for the specified repository and tag."""
+    branches = repository.get_branches()
+    matched_branches = [match for match in branches if match.name == branch]
```

Comment:
`repository.get_branch` can simplify all of this, but I am not sure it is even needed. Tags should def be used instead of branches, so you can instead use `get_tag`

Reply:
So I removed both `get_sha_for_branch` and `get_integrations_package`. This code was consolidated into `get_integration_manifests` and `build_integrations_manifest`.

The github call authentication and prod package-storage setup is now done in `build_integrations_manifest` as it does not need to be done for every integration, rather once.

Regarding the sha vs tag, package-storage only has 1 tag that dates back to 2020. This is why we chose to pull the commit sha for the latest, specific to each package integration in production. This still has the expectation that if a package exists in the production branch, it started from snapshot and was promoted after passing tests.